File: Exame2_CS/POO/POO_4.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Exame2_CS.POO
{
    public class Z
    {
        public void M() { Console.WriteLine("Z.M"); }
        public virtual void P() { Console.WriteLine("Z.P"); }
    }

    public class Y : Z
    {
        public void M() { Console.WriteLine("Y.M"); }
        public override void P() { Console.WriteLine("Y.P"); }
    }

    public class X : Y
    {
        public void M() { Console.WriteLine("X.M"); }
    }

    public class W : X
    {

        public void M() { Console.WriteLine("W.M"); }
    }


    public class POO_4
    {
        public static void Exec()
        {
            W d = new W();
            Z a = d;
            Y b = d;
            X c = d;

            a.M();
            b.M();
            c.M();
            d.M();

            Y b1 = new Y();
            Z a1 = b;

            a1.P();
            b1.P();

            Console.ReadLine();
        }
    }
}

File: Examen2/POO/POO_4.vb
Class Z
    Public Sub M()
        Console.WriteLine("Z.M")
    End Sub
    Public Overridable Sub P()
        Console.WriteLine("Z.P")
    End Sub
End Class

Class Y : Inherits Z
    Public Sub M()
        Console.WriteLine("Y.M")
    End Sub
    Public Overrides Sub P()
        Console.WriteLine("Y.P")
    End Sub
End Class

Class X : Inherits Y
    Public Sub M()
        Console.WriteLine("X.M")
    End Sub
End Class

Class W : Inherits X
    Public Sub M()
        Console.WriteLine("W.M")
    End Sub
End Class


Public Class POO_4
    Public Shared Sub Exec()
        Dim d As W = New W
        Dim a As Z = d
        Dim b As Y = d
        Dim c As X = d

        a.M()
        b.M()
        c.M()
        d.M()

        Dim b1 As Y = New Y
        Dim a1 As Z = b
        a1.P()
        b1.P()

        Console.ReadLine()
    End Sub
End Class